Destroyed module leaves config behind in state

### Terraform Version

```
Terraform v0.12.7
+ provider.aws v2.11.0
+ provider.bigip v0.12.3
+ provider.chef v0.2.0
+ provider.consul v2.5.0
+ provider.null v2.1.2
+ provider.template v2.1.2
+ provider.vsphere v1.12.0
```

### Terraform Configuration Files

I have a project that utilizes a module for creating a vsphere instances. I had two roles and thus two blocks for this module. I removed one of the blocks to destroy those instances. The removed block looked like this:

```hcl
module "vsphere_linux_master" {
  source = "../../modules/vsphere_instance_linux_ng"

  instance_count = lower(var.env_type) == "qa" ? "1" : "0"

  providers = {
    aws     = aws
    vsphere = vsphere.config
  }

  environment      = lower(var.env_type)
  product          = lower(var.product)
  application      = lower(var.application)
  component        = "master"
  region           = "sac"
  service_provider = "op"

  vsphere_username      = var.vsphere_username
  vsphere_password      = var.vsphere_password
  vsphere_cluster       = var.vsphere_cluster
  vsphere_datastore     = var.vsphere_datastore
  vsphere_folder_path   = var.vsphere_folder_path
  vsphere_datacenter    = var.vsphere_datacenter
  vsphere_network_label = var.vsphere_network_label

  vsphere_vcpu   = var.vsphere_vcpu
  vsphere_memory = var.vsphere_memory
  data_disk_size = var.data_disk_size
}
```

The destroy apply went fine and now about a month later I go to do work in the project again and I find that during an apply, I get this error *after the apply otherwise successfully goes through*:

```
...
module.coredns.aws_instance.instance[1]: Modifying... [id=i-1234]
module.coredns.aws_instance.instance[0]: Modifying... [id=i-1235]
module.coredns.aws_instance.instance[1]: Modifications complete after 3s [id=i-1234]
module.coredns.aws_instance.instance[0]: Modifications complete after 3s [id=i-1235]

Error: Missing required argument

The argument "user" is required, but was not set.
```

I looked at the trace below and saw that it's still parsing provider config for the removed module resources. So I looked at the state and saw that indeed there are remnants of the module still in the state.

```
    {
      "module": "module.vsphere_linux_master",
      "mode": "managed",
      "type": "vsphere_compute_cluster_vm_anti_affinity_rule",
      "name": "cluster_vm_anti_affinity_rule",
      "each": "list",
      "provider": "provider.vsphere.config",
      "instances": []
    },
...
    {
      "module": "module.vsphere_linux_master",
      "mode": "managed",
      "type": "vsphere_virtual_disk",
      "name": "data_disk",
      "each": "list",
      "provider": "provider.vsphere.config",
      "instances": []
    },
...
```

After removing these sections from the state manually, the error no longer occurs. Note that the error only happens during an `apply`, not during a `plan`.

### Debug Output
Relevant section of a trace output:
```
2019/09/10 09:06:59 [TRACE] <root>: eval: *terraform.EvalSetModuleCallArguments
2019/09/10 09:06:59 [TRACE] [walkApply] Exiting eval tree: module.coredns.var.security_group_ids
2019/09/10 09:06:59 [TRACE] vertex "module.coredns.var.security_group_ids": visit complete
2019/09/10 09:06:59 [WARN] <root>: eval: *terraform.EvalConfigProvider, non-fatal err: Missing required argument: The argument "user" is required, but was not set.
2019/09/10 09:06:59 [ERROR] <root>: eval: *terraform.EvalSequence, err: Missing required argument: The argument "user" is required, but was not set.
2019/09/10 09:06:59 [ERROR] <root>: eval: *terraform.EvalOpFilter, err: Missing required argument: The argument "user" is required, but was not set.
2019/09/10 09:06:59 [ERROR] <root>: eval: *terraform.EvalSequence, err: Missing required argument: The argument "user" is required, but was not set.
2019/09/10 09:06:59 [TRACE] [walkApply] Exiting eval tree: provider.vsphere
2019/09/10 09:06:59 [TRACE] vertex "provider.vsphere": visit complete
2019/09/10 09:06:59 [TRACE] dag/walk: upstream of "module.vsphere_linux_master.vsphere_virtual_disk.data_disk (clean up state)" errored, so skipping
2019/09/10 09:06:59 [TRACE] dag/walk: upstream of "module.vsphere_linux_master.vsphere_virtual_machine.instance (clean up state)" errored, so skipping
2019/09/10 09:06:59 [TRACE] dag/walk: upstream of "module.vsphere_linux_master.vsphere_compute_cluster_vm_anti_affinity_rule.cluster_vm_anti_affinity_rule (clean up state)" errored, so skipping
2019/09/10 09:06:59 [TRACE] dag/walk: upstream of "module.vsphere_linux_master (removed)" errored, so skipping
2019/09/10 09:06:59 [TRACE] dag/walk: upstream of "provider.vsphere (close)" errored, so skipping
```

### Expected Behavior

Destroying the module resources by removing the module block should have completely removed all elements from the state.

### Actual Behavior

State contained blocks for the destroyed module and caused a provider config error.

### Steps to Reproduce

1. Instantiate a module that accepts config for a provider.
2. Destroy the module instances by removing its code block.
3. Observe remaining elements in state.

### Additional Context

A `terraform state list` does not show anything regarding the destroyed module, probably because the entry's instances are empty in the state. But terraform is still clearly trying to parse the provider config for the destroyed module during an `apply`.  

The destroy operations were I believe performed with terraform 0.12.4.

--- a/terraform/graph_builder_apply_test.go
+++ b/terraform/graph_builder_apply_test.go
@@ -600,6 +600,48 @@ test_object.b
 	}
 }
 
+// The orphan clean up node should not be connected to a provider
+func TestApplyGraphBuilder_orphanedWithProvider(t *testing.T) {
+	changes := &plans.Changes{
+		Resources: []*plans.ResourceInstanceChangeSrc{
+			{
+				Addr: mustResourceInstanceAddr("test_object.A"),
+				ChangeSrc: plans.ChangeSrc{
+					Action: plans.Delete,
+				},
+			},
+		},
+	}
+
+	state := states.NewState()
+	root := state.EnsureModule(addrs.RootModuleInstance)
+	root.SetResourceInstanceCurrent(
+		mustResourceInstanceAddr("test_object.A").Resource,
+		&states.ResourceInstanceObjectSrc{
+			Status:    states.ObjectReady,
+			AttrsJSON: []byte(`{"id":"A"}`),
+		},
+		mustProviderConfig("provider.test.foo"),
+	)
+
+	b := &ApplyGraphBuilder{
+		Config:     testModule(t, "graph-builder-orphan-alias"),
+		Changes:    changes,
+		Components: simpleMockComponentFactory(),
+		Schemas:    simpleTestSchemas(),
+		State:      state,
+	}
+
+	g, err := b.Build(addrs.RootModuleInstance)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// The cleanup node has no state or config of its own, so would create a
+	// default provider which we don't want.
+	testGraphNotContains(t, g, "provider.test")
+}
+
 const testApplyGraphBuilderStr = `
 meta.count-boundary (EachMode fixup)
   module.child.test_object.other

--- a/terraform/node_resource_destroy.go
+++ b/terraform/node_resource_destroy.go
@@ -277,36 +277,15 @@ func (n *NodeDestroyResourceInstance) EvalTree() EvalNode {
 // leaving skeleton resource objects in state after their instances have
 // all been destroyed.
 type NodeDestroyResource struct {
-	*NodeAbstractResource
+	NodeAbstractResource *NodeAbstractResource
 }
 
 var (
-	_ GraphNodeResource      = (*NodeDestroyResource)(nil)
-	_ GraphNodeReferenceable = (*NodeDestroyResource)(nil)
-	_ GraphNodeReferencer    = (*NodeDestroyResource)(nil)
-	_ GraphNodeEvalable      = (*NodeDestroyResource)(nil)
+	_ GraphNodeEvalable = (*NodeDestroyResource)(nil)
 )
 
 func (n *NodeDestroyResource) Name() string {
-	return n.ResourceAddr().String() + " (clean up state)"
-}
-
-// GraphNodeReferenceable, overriding NodeAbstractResource
-func (n *NodeDestroyResource) ReferenceableAddrs() []addrs.Referenceable {
-	// NodeDestroyResource doesn't participate in references: the graph
-	// builder that created it should ensure directly that it already depends
-	// on every other node related to its resource, without relying on
-	// references.
-	return nil
-}
-
-// GraphNodeReferencer, overriding NodeAbstractResource
-func (n *NodeDestroyResource) References() []*addrs.Reference {
-	// NodeDestroyResource doesn't participate in references: the graph
-	// builder that created it should ensure directly that it already depends
-	// on every other node related to its resource, without relying on
-	// references.
-	return nil
+	return n.NodeAbstractResource.ResourceAddr().String() + " (clean up state)"
 }
 
 // GraphNodeEvalable
@@ -316,6 +295,6 @@ func (n *NodeDestroyResource) EvalTree() EvalNode {
 	// leftover husk of a resource in state after all of the child instances
 	// and their objects were destroyed.
 	return &EvalForgetResourceState{
-		Addr: n.ResourceAddr().Resource,
+		Addr: n.NodeAbstractResource.ResourceAddr().Resource,
 	}
 }

--- a/terraform/testdata/graph-builder-orphan-alias/main.tf
+++ b/terraform/testdata/graph-builder-orphan-alias/main.tf
@@ -1,0 +1,3 @@
+provider "test" {
+  alias = "foo"
+}

